### PR TITLE
Figure out add-server, remove-server, and add some specs for them

### PR DESCRIPTION
**Pull request by cap10morgan**
Once I figured out how to use add-server and remove-server I realized that they were already working but it wasn't very obvious what the shape of the arguments' data needed to be. This seemed like a great use case for clojure.spec, so I added some to hopefully make it clearer what these fns need next time someone is exploring this library.

I also added a `kv-example/run` fn and a `make kv-example` target that calls it. This runs the kv-example through some server additions and removals (including the leader at one point) all while updating an "index" key in the kv data that the raft network is consensing around.

The `add-server` and `remove-server` ops seem to be doing the right things according to the raft paper and this update to it: https://www.openlife.cc/sites/default/files/4-modifications-for-Raft-consensus.pdf. But we may want to add the `databaseId` concept from the latter to our implementation (though perhaps with a different name? kind of confusing in a library that is used by a database).

I also applied some opportunistic refactors & improvements since I was in here. Hopefully those are more or less isolated to their own commits.

I encourage pulling this branch down and running `make kv-example` and tailing `kv-example.log` (I recommend running `less +F -R kv-example.log` to see the formatting in the log correctly) as a part of reviewing this PR. ~~I'm going to try using `raft/add-server` and `raft/remove-server` in ledger to help support dynamic cluster sizing.~~

UPDATE: We're not going to try to support dynamic raft clusters after all. It probably only really makes sense to have 3 or 5 member raft clusters, and they should just be statically defined at startup.

### CODE REVIEW

**Comment by cap10morgan** [comment]
This is all related to FC-332 and FC-1107.

File: src/fluree/raft.clj [hunk 1]
```diff
@@ -52,20 +58,49 @@
...
       :closed)))
 
 
+(defn add-server
+  "Adds an already created server to the raft network. You can create the
+  server by calling `start` with the appropriate config, then pass its id to
+  this fn in the `new-server-id` arg. Due to the way the raft protocol works,
+  you can only add one server at a time."
+  [raft new-server-id]
+  (raft-specs/valid-or-throw ::raft-specs/raft raft "invalid raft state")
```
Comment (bplatz) [inline src/fluree/raft.clj:83]: I think we might be better off with an assertion here as I don't think user input would change this. Assuming this is primarily for development validation, an assertion would catch errors during development but not unnecessarily slow down production.

File: src/fluree/raft/kv_example.clj [hunk 2]
```diff
@@ -199,9 +199,15 @@
...
-                                                                              (:leader x) " term: " (:term x))))
+                                        :leader-change-fn      (fn [{:keys [event cause message old-leader
+                                                                            new-leader new-raft-state]}]
+                                                                 (log/info
+                                                                   (format "%s reports leader change from %s to %s "
+                                                                           "term: %s event: %s cause: %s"
+                                                                           "message: %s"
+                                                                           server-id old-leader new-leader
+                                                                           (:term new-raft-state)
+                                                                           event cause message)))
```
Comment (bplatz) [inline src/fluree/raft/kv_example.clj:214]: 🥇

**Review by bplatz — APPROVED** [review]
Looks good, note comment if applicable.

File: src/fluree/raft.clj (hunk 1 above)
Comment (cap10morgan) [reply src/fluree/raft.clj:83]: Yeah, I went back and forth on this too. I wouldn't want to call `valid-or-throw` anywhere that is invoked often or in a very performance-critical section of code. I figured `add-server` was sufficiently seldom invoked and not performance sensitive enough for this to make a noticeable difference. But what do you think?

If we do start adding these checks anywhere that would negatively impact performance in production, then I agree that we should use assertions we turn off in production.

Comment (bplatz) [reply src/fluree/raft.clj:83]: Agree probably not an issue here. Concern would be less about frequency of calls and more on Raft timeouts if this operation is blocking a broadcast message (synchronous inside the heartbeat). With a sufficiently low heartbeat timeout, a millisecond can make a difference.

Comment (cap10morgan) [reply src/fluree/raft.clj:83]: Makes sense. I'll investigate a bit before merging.

**Comment by cap10morgan** [comment]
Not ready for merging just yet. I need to test performance impact of the spec checking.

**Pull request closed without merging**